rfcs/20221010-add-ocl-ooo-queue/README.md

```diff
@@ -0,0 +1,125 @@
+supports flags that controls stream behavior. The flags include `in_order` and `out_of_order`
+values that can be used to create an out-of-order stream. Currently, oneDNN returns a status
+`unimplemented` when a user tries to create an out-of-order stream for OpenCL runtime
+therefore no new API is required. The out-of-order stream will be made available for creation
+in the case of OpenCL runtime.
+
+### OpenCL Interop API
+The API is an extension of the generic execute API and has two additional responsibilities:
+1. Take a vector of events for dependencies
+2. Return an event for the primitive that was submitted for execution
+
+#### C API
+
+```c
+// dnnl_ocl.h
+
+/// Executes computations specified by the primitive in a specified stream and
+/// returns an OpenCL event.
+///
+/// @param primitive Primitive to execute.
+/// @param stream Stream to use.
+/// @param nargs Number of arguments.
+/// @param args Array of arguments. Each argument is an
+///     <index, #dnnl_memory_t> pair. The index is one of the `DNNL_ARG_*`
+///     values such as `DNNL_ARG_SRC`. Unless runtime shapes are used (see
+///     #DNNL_RUNTIME_DIM_VAL), the memory object must have the same memory
+///     descriptor as that returned by
+///     #dnnl_primitive_desc_query_md(#dnnl_query_exec_arg_md, index).
+/// @param deps A pointer to std::vector<cl_event> that contains
+///     dependencies.
+/// @param return_event Output event. The library doesn't own the event.
+///     It is the user's responsibility to manage lifetime of the event.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_ocl_interop_primitive_execute(
+        const_dnnl_primitive_t primitive, dnnl_stream_t stream, int nargs,
+        const dnnl_exec_arg_t *args, const cl_event *deps, int ndeps,
+        cl_event *return_event);
+```
+
+#### C++ API
+```cpp
+// dnnl_ocl.hpp
+
+/// Executes computations specified by the primitive in a specified stream and
+/// returns an OpenCL event.
+///
+/// Arguments are passed via an arguments map containing
+/// <index, memory object> pairs. The index must be one of the `DNNL_ARG_*`
+/// values such as `DNNL_ARG_SRC`, and the memory must have a memory descriptor
+/// matching the one returned by
+/// #dnnl::primitive_desc::query_md(#query::exec_arg_md, index) unless using
+/// dynamic shapes (see #DNNL_RUNTIME_DIM_VAL).
+///
+/// @param aprimitive Primitive to execute.
+/// @param astream Stream object. The stream must belong to the same engine
+///     as the primitive.
+/// @param args Arguments map.
+/// @param deps Optional vector with `cl::sycl::event` dependencies.
+///
+/// @returns Output event. The library doesn't own the event. It is the user's
+///     responsibility to manage lifetime of the event.
+inline cl_event execute(const dnnl::primitive &aprimitive,
+        const stream &astream, const std::unordered_map<int, memory> &args,
+        const std::vector<cl_event> &deps = {});
```

Comment:
Have you considered adding a wrapper for events and having common API for different backends with downcast to cl/sycl event internally? If we add support for SYCL/L0 backends in OV, it would be simpler to add a conversion from our event object to onednn event and have exactly the same API calls.

Reply:
> Have you considered adding a wrapper for events and having common API for different backends

We have considered that in the very beginning but the value of having such an abstraction is questionable. 
In the case of dependencies (input events) you would need to create the oneDNN event abstraction for the given OpenCL/SYCL event. For the output event you would need to query an encapsulated OpenCL/SYCL event to use it with OpenCL/SYCL API. So the abstraction has little to no responsibilities. 

> it would be simpler to add a conversion from our event object to onednn event and have exactly the same API calls.

The API for creating oneDNN event would still be in backend specific interop namespaces because it would take an OpenCL or SYCL event. So conversion from an OpenVINO event to oneDNN event would look very similar to calling OpenCL and SYCL execute functions.